cloud: add random access support to ResumingReader

sravotto opened the pull request:
This commit adds ReadAt and Seek implementations to ResumingReader and its wrapper types (limitedReader, metricsReader), enabling random access to cloud-stored files.

Implementation details:
- ReadAt creates a new reader at the specified offset using the Opener
- Checks bounds against known file size before opening to avoid invalid
  range requests to cloud storage
- Uses cached Size field for efficient bounds checking on subsequent calls
- Seek tracks position logically without seeking the underlying reader
- Both methods implement the new RandomAccessFile interface
- SupportsRandomAccess() helper function for clean type checking
- Rate limiting uses batching for both Read() and ReadAt() to reduce
  overhead (128KB batches shared via l.pool)

This enables efficient reading of formats like Parquet that require random access, without buffering entire files to disk.

Release note: none

Epic: CRDB-23802

sravotto committed:
cloud: add random access support to ResumingReader

This commit adds ReadAt and Seek implementations to ResumingReader and
its wrapper types (limitedReader, metricsReader), enabling random access
to cloud-stored files.

Implementation details:
- ReadAt creates a new reader at the specified offset using the Opener
- Checks bounds against known file size before opening to avoid invalid
  range requests to cloud storage
- Uses cached Size field for efficient bounds checking on subsequent calls
- Seek tracks position logically without seeking the underlying reader
- Both methods implement the new RandomAccessFile interface
- SupportsRandomAccess() helper function for clean type checking
- Rate limiting: Read() uses batching (128KB pool), ReadAt() calls rate
  limiter directly to maintain thread-safety per io.ReaderAt contract

This enables efficient reading of formats like Parquet that require
random access, without buffering entire files to disk.

Release note: none

Epic: CRDB-23802

jeffswenson commented on pkg/cloud/cloud_io.go:
```diff
@@ -378,6 +380,57 @@ func (r *ResumingReader) Close(ctx context.Context) error {
 
+// ReadAt implements ioctx.ReaderAtCtx using the Opener to create a reader at the specified offset.
+// This allows formats like Parquet that require random access to work efficiently with cloud storage.
+func (r *ResumingReader) ReadAt(ctx context.Context, p []byte, off int64) (n int, err error) {
```
We aren't really using the ResumingReader here, we are basically using it as a wrapper for ReaderOpenerAt. I think we should make this its own utility instead of adding it on to ResumingReader. The parquet import implementation can use this adapter without modifying the cloud package.

```go
type readAt struct {
     storage cloud.ExternalStorage
     basename string
}

func (r *ReadAt) ReadAt(ctx context.Context, p []byte, off int64) (n int, err error) {
    f, err := r.ReadFile(ctx, r.basename, cloud.ReadOptions {
        Offset: off,        // read from the right offset
        LenghtHint: len(p), // avoid buffering more bytes in the cloud object than are needed
        NoFileSize: true,   // avoid a round trip on some platforms
    })
    if err { return 0, err }
    defer reader.Close()
    return ioctx.ReadFull(reader, p)
}
```

This also lets us avoid the changes to the limit and metric readers. So the overall change gets smaller.

jeffswenson commented on pkg/cloud/metrics.go:
```diff
@@ -186,6 +187,24 @@ func (mr *metricsReader) Close(ctx context.Context) error {
+// This allows Parquet and other formats that require random access to work with
+// cloud storage backends without buffering the entire file to disk.
+func (mr *metricsReader) ReadAt(ctx context.Context, p []byte, off int64) (n int, err error) {
+	if readerAt, ok := mr.ReadCloserCtx.(ioctx.ReaderAtCtx); ok {
```
I think we can avoid these casts by defining a `cloud.ReadFile` interface that is:

```
type ReadFile interface {
    ioctx.ReadCloserCtx
    ioctx.ReaderAtCtx
    ioctx.SeekerCtx
}
```

Then adjusting ExternalStorage.ReadFile to return it instead of ioctx.ReadCloserCtx. That also ensures that all of the cloud storage implementations will be usable with parquet import.

sravotto replied:
I created a `ReadFile` interface as suggested, however I added a separate `RandomReadFile` interface to make random access support explicit.
<!-- Reviewable comment -OfWkHB-8YnpB-9Qw0hx:b8tijnr -->
<!-- Sent from Reviewable.io -->

sravotto reviewed (commented):
Thanks for the review, please take another look.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158209#-:-Oevjp1ODh9wTivEYgcN:bf2248v)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @BramGruneir, @jeffswenson, and @mw5h)



<!-- Sent from Reviewable.io -->

mw5h reviewed (changes requested):
> @mw5h reviewed 1 of 4 files at r1, 2 of 4 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158209)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @BramGruneir, @jeffswenson, and @sravotto)

___
*[`pkg/cloud/cloud_io.go` line 396 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/158209#-OfQBUrbCkGpGMHJ1oU3:-OfQBUrbCkGpGMHJ1oU4:bid0df3) ([raw file](https://github.com/cockroachdb/cockroach/blob/985e2a8e3f9023a33e91d8fd07f3fc63cbd42f08/pkg/cloud/cloud_io.go#L396)):*
> ```Go
> 		}
> 		// Truncate read to available bytes
> 		if off+int64(len(p)) > r.Size {
> ```

Won't this cause the read to fail with ErrUnexpectedEOF below?


<!-- Sent from Reviewable.io -->

sravotto committed:
cloud: refactor ReadFile interface and fix ReadAt EOF handling

This commit makes several improvements to the cloud storage I/O interfaces:

1. Split ReadFile interface into base and extended versions:
   - ReadFile: base interface for sequential reads
   - RandomReadFile: extends ReadFile with ReadAt/Seek for random access

2. Fixed ResumingReader.ReadAt to comply with io.ReaderAt semantics:
   - Return io.EOF when reading past end-of-file

3. Updated all ExternalStorage implementations and wrappers:
   - Changed ReadFile return type from ioctx.ReadCloserCtx to cloud.ReadFile
   - Updated test mocks (fakeExternalStorage, generatorExternalStorage,
     mockSinkStorage)
   - Fixed metricsReader and limitedReader to properly delegate random access
   - Added ReadAt/Seek support to accountingReader for tenant accounting

Release note: none

Epic: CRDB-23802

sravotto reviewed (commented):
Adding the `ReadFile` interface required to modify several packages. I also discovered that the `multitenantio.accountingReader` needed to be changed to support `ReadAt/Seek`.  Maintaining thread safety for `ReadAt` makes the code a little more complicated.

I decided to have a separate `RandomReadFile` interface to make random access support explicit. The alternative (single unified interface) would require creating wrapper/adapters for each non-random-access implementation to return `ErrRandomAccessNotSupported` for `ReadAt/Seek`. I'm open to implement that, but I am thinking that the separate `RandomReadFile` seems to follow Go's interface composition pattern (similar to
  io.Reader vs io.ReadSeeker in the standard library). 

Note: New changes are in a separate commit.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158209#-:-OfWo2Rz31BJJ19VV9XA:b-mon492)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @BramGruneir, @jeffswenson, @KeithCh, @mgartner, and @mw5h)

___
*[`pkg/cloud/cloud_io.go` line 396 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/158209#-OfQBUrbCkGpGMHJ1oU3:-OfWjopG3w_VgAZUFgsW:bbf85jo) ([raw file](https://github.com/cockroachdb/cockroach/blob/985e2a8e3f9023a33e91d8fd07f3fc63cbd42f08/pkg/cloud/cloud_io.go#L396)):*
<details><summary><i>Previously, mw5h (Matt White) wrote…</i></summary><blockquote>

Won't this cause the read to fail with ErrUnexpectedEOF below?
</blockquote></details>

I refactored it a bit. I did want to retain the ReadAt semantics where you get an error (which I changed to io.EOF) along with the bytes read, in case we reach EOF before filling the buffer.


<!-- Sent from Reviewable.io -->

sravotto commented:
The approach suggested  by @jeffswenson  is simpler and the work is better isolated.

Closing in favor of #159067.
